src/data/string/ustring_utf_8_decode_test.cpp

```diff
@@ -0,0 +1,174 @@
+// Copyright (c)2008-2011, Preferred Infrastructure Inc.
+// 
+// All rights reserved.
+// 
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+// 
+//     * Redistributions of source code must retain the above copyright
+//       notice, this list of conditions and the following disclaimer.
+// 
+//     * Redistributions in binary form must reproduce the above
+//       copyright notice, this list of conditions and the following
+//       disclaimer in the documentation and/or other materials provided
+//       with the distribution.
+// 
+//     * Neither the name of Preferred Infrastructure nor the names of other
+//       contributors may be used to endorse or promote products derived
+//       from this software without specific prior written permission.
+// 
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include <gtest/gtest.h>
+
+#include "./ustring.h"
+
+using namespace pfi::data::string;
+
+TEST(ustring_utf_8_decode_test, utf_8_byte_sequences_are_empty)
+{  const char* p="";
+  EXPECT_THROW({chars_to_uchar(p, p + strlen(p));}, std::invalid_argument);
+}
+
+TEST(ustring_utf_8_decode_test, utf_8_single_byte_character_is_out_of_range)
+{
+  for (int c = 0x80; c < 0xBF; c++) {
+    std::string str(1u, static_cast<char>(c));
+    const char* p=str.c_str();
+    EXPECT_THROW({chars_to_uchar(p, p + str.size());}, std::invalid_argument)
+      << std::hex << c;
+  }
+  for (int c = 0xFE; c <= 0xFF; c++) {
+    std::string str(1u, static_cast<char>(c));
+    const char* p=str.c_str();
+    EXPECT_THROW({chars_to_uchar(p, p + str.size());}, std::invalid_argument)
+      << std::hex << c;
+  }
+}
+
+TEST(ustring_utf_8_decode_test, utf_8_byte_sequences_have_a_5_byte_sequence)
+{
+  { // min 5-byte sequence
+    const char str[] = {0xF8, 0x80, 0x80, 0x80, 0x80, 0x00};
+    const char* p=str;
+    EXPECT_THROW({chars_to_uchar(p, p + strlen(p));}, std::invalid_argument) << str;
```

Comment:
Since `str` is invalid UTF-8 string, `<< str` will cause `unittest_gtest.py` to crash when the test failed, as below
```
  tests that fail 1/23
    .../pficommon/build/src/data/ustring_utf_8_decode_test
Traceback (most recent call last):
  File ".../pficommon/.waf-1.7.8-77e2cdd38ea2930f982e3aabf335fdb7/waflib/Scripting.py", line 97, in waf_entry_point
    run_commands()
  File ".../pficommon/.waf-1.7.8-77e2cdd38ea2930f982e3aabf335fdb7/waflib/Scripting.py", line 153, in run_commands
    ctx=run_command(cmd_name)
  File ".../pficommon/.waf-1.7.8-77e2cdd38ea2930f982e3aabf335fdb7/waflib/Scripting.py", line 146, in run_command
    ctx.execute()
  File ".../pficommon/.waf-1.7.8-77e2cdd38ea2930f982e3aabf335fdb7/waflib/Scripting.py", line 351, in execute
    return execute_method(self)
  File ".../pficommon/.waf-1.7.8-77e2cdd38ea2930f982e3aabf335fdb7/waflib/Build.py", line 106, in execute
    self.execute_build()
  File ".../pficommon/.waf-1.7.8-77e2cdd38ea2930f982e3aabf335fdb7/waflib/Build.py", line 124, in execute_build
    self.post_build()
  File ".../pficommon/.waf-1.7.8-77e2cdd38ea2930f982e3aabf335fdb7/waflib/Build.py", line 287, in post_build
    m(self)
  File ".../pficommon/unittest_gtest.py", line 272, in summary
    print(out.decode('utf-8'))
  File "/usr/lib64/python2.7/encodings/utf_8.py", line 16, in decode
    return codecs.utf_8_decode(input, errors, True)
UnicodeDecodeError: 'utf8' codec can't decode bytes in position 1721-1723: invalid continuation byte
```

Even if python can handle invalid UTF-8 string, we will get `�` (U+FFFD) or something strange sequence.
Unpacking `str` into hex or no `<< str` is better?